### DISCUSSION TITLE
Make function run before all fixtures

### DISCUSSION THREAD

**Discussion by andyhasit** [pytest-dev/pytest#10395] [General] [answered]
I want to conditionally code for some tests _before_ the test's other fixtures run. Its to execute SQL to toggle logging on or off, which needs to happen before the other fixtures create records in the DB.

I ideally want to inspect the test's name to determine whether to toggle logging, alternatively I'm happy to use a fixture.

* I can't seem to do this using a fixture because I have no way of guaranteeing it would run before the other fixtures, which vary in scope, and I'm not able to go and change those scopes. The only way I can see of creating a fixture that runs before all other is using `(scope="session", autouse=True)` but I need it to run right before each test, so function scoped.
* I can't seem to do this using any of the `pytest_runtest_*` hooks as the fixtures run before those.

Surely I'm just not finding a builtin hook that runs at the point I need it? I've spent hours poring over the docs and stackoverflow and still can't find it.

**Comment by andyhasit** [comment, marked as answer]
Sorry, I just realised what I'm asking for is not possible, as a test could require a fixture of higher scope than function. 

I'll just need to create a function-scoped fixture which sends the statements and require that in other fixtures I use on those tests, which would force them to be function-scoped too.

I think it's either that or separating it into two runs with mutually exclusive ignore and exclude patterns which isn't terrible but is really not great either.

**Comment by RonnyPfannschmidt** [comment]
@andyhasit 
https://github.com/pytest-dev/pytest/blob/e54c6a1362589b32a2e63bb780192b86216ecec8/src/_pytest/runner.py#L153-L155 

impleemnts te setup

even high scope scopes - so having that hook is perfectly fine to implement it - however you need a @hookimpl(tryfirst=true)

**Reply by andyhasit** [reply]
@RonnyPfannschmidt I believe you meant something like this:

```python
@pytest.hookimpl(tryfirst=True)
def pytest_runtest_setup(item):
    if "audit" in item.name:
        print("enable db logging")
    else:
        print("disable db logging")
```

However that does not work for my stated aim, which is:

> to execute SQL to toggle logging on or off, which needs to happen before the other fixtures create records in the DB.

Consider the following:

```python
@pytest.fixture(scope="session")
def fixt_a():
    print('create db object a')

@pytest.fixture
def fixt_b():
    print('create db object b')

def test_normal(fixt_a, fixt_b):
    print('test_normal')

def test_with_audit(fixt_a, fixt_b):
    print('test_with_audit')
``` 

Which produces the following output:

```
disable db logging
create db object a
create db object b
test_normal
enable db logging
create db object b
test_with_audit
```

Because **fixture a** is session scoped, it only runs once, meaning **object a** will only be created once, while logging is disabled, and **test_with_audit** will not find log record for **object a**.

The resolution to this problem is, as I stated in my answer, that what I'm asking for is not possible, and the only solution is to either have two completely separate runs, or to use a fixture and ensure it is always of a higher scope than any fixtures which create db records (although I may have missed some options).

Additionally, I need to disable logging after running those tests which need it enabled. I settled on module scope as it's easy to group tests which need this into modules.

```python
@pytest.fixture(scope="module")
def enable_audit_trigger(test_db):
    test_db.enable_triggers()
    yield
    test_db.disable_triggers()
```

I think there is still a hole in my plan: I currently have no way of ensuring that tests which require this fixture don't also require fixtures which create database records at a higher scope, or at module scope but run first.
